baseline/tf/classify/training/utils.py

```diff
@@ -252,3 +253,85 @@ def recover_last_checkpoint(self):
+        """Create a Trainer, and give it the parameters needed to instantiate the model
+
+        :param model_params: The model parameters
+        :param kwargs: See below
+
+        :Keyword Arguments:
+
+          * *nsteps* (`int`) -- If we should report every n-steps, this should be passed
+          * *ema_decay* (`float`) -- If we are doing an exponential moving average, what decay to us4e
+          * *clip* (`int`) -- If we are doing gradient clipping, what value to use
+          * *optim* (`str`) -- The name of the optimizer we are using
+          * *lr* (`float`) -- The learning rate we are using
+          * *mom* (`float`) -- If we are using SGD, what value to use for momentum
+          * *beta1* (`float`) -- Adam-specific hyper-param, defaults to `0.9`
+          * *beta2* (`float`) -- Adam-specific hyper-param, defaults to `0.999`
+          * *epsilon* (`float`) -- Adam-specific hyper-param, defaults to `1e-8
+
+        """
+        super().__init__(model_params, **kwargs)
+        self.threshold = float(kwargs.get("output_threshold", 0.5))
+
+    def _test(self, loader, **kwargs):
+        """Test an epoch of data using either the input loader or using `tf.dataset`
+
+        In non-`tf.dataset` mode, we cycle the loader data feed, and pull a batch and feed it to the feed dict
+        When we use `tf.dataset`s under the hood, this function simply uses the loader to know how many steps
+        to train.
+
+        :param loader: A data feed
+        :param kwargs: See below
+
+        :Keyword Arguments:
+          * *dataset* (`bool`) Set to `True` if using `tf.dataset`s, defaults to `True`
+          * *reporting_fns* (`list`) A list of reporting hooks to use
+          * *verbose* (`dict`) A dictionary containing `console` boolean and `file` name if on
+
+        :return: Metrics
+        """
+        if self.ema:
+            self.sess.run(self.ema_load)
+
+        use_dataset = kwargs.get('dataset', True)
+
+        gold_labels = []
+        pred_labels = []
+        steps = len(loader)
+        total_loss = 0
+        total_norm = 0
+        verbose = kwargs.get("verbose", None)
+
+        pg = create_progress_bar(steps)
+        for i, batch_dict in enumerate(pg(loader)):
+            y = batch_dict['y']
+            if use_dataset:
+                preds, lossv = self.sess.run([self.model.probs, self.test_loss])
+            else:
+                feed_dict = self.model.make_input(batch_dict, False)
+                preds, lossv = self.sess.run([self.model.probs, self.test_loss], feed_dict=feed_dict)
+
+            batchsz = len(preds)
+            total_loss += lossv * batchsz
+            total_norm += batchsz
+
+            gold_labels.append(y.astype(np.int32))
+            preds[preds > self.threshold] = 1
+            pred_labels.append(preds.astype(np.int32))
+
+        metrics = ml.get_all_metrics(np.concatenate(gold_labels, axis=0), np.concatenate(pred_labels, axis=0))
```

Comment:
seems like overkill that we should have to redefine the entire training loop to override the metrics

Reply:
additionally, as we fully support eager mode now (and will probably start recommending it over declarative), there needs to be a full solution for eager and eager-distributed

Reply:
I'll look into added eager stuff, I always forget it about it lol

Reply:
> seems like overkill that we should have to redefine the entire training loop to override the metrics

I agree, what do we think is the best solution? There seems to be a couple ways.

1) Conditional execution based on if the model is multilabel or not.
2) Add hooks that make swapping evaluation functions easier, The training functions have always seemed a little large and hard to customize to me and it seems like this could be a chance to make it more configurable and maybe share things better
3) Overrides like this. (Which we decided was bad lol)

Also one consideration if how this effects future use of things like tf.functions which you mentioned really sped up the taggers.